tools/bbc/bbc.cpp

```diff
@@ -0,0 +1,204 @@
+
+// Convert Fortran input to MLIR (target is FIR dialect)
+void convertFortranSourceToMLIR(
+    std::string path, Fortran::parser::Options options, DriverOptions &driver,
+    Fortran::semantics::SemanticsContext &semanticsContext) {
+  if (!driver.forcedForm) {
+    auto dot{path.rfind(".")};
+    if (dot != std::string::npos) {
+      std::string suffix{path.substr(dot + 1)};
+      options.isFixedForm = suffix == "f" || suffix == "F" || suffix == "ff";
+    }
+  }
+  options.searchDirectories = IncludeDirs;
+  Fortran::parser::Parsing parsing{semanticsContext.allSources()};
+  parsing.Prescan(path, options);
+  if (!parsing.messages().empty() &&
+      (driver.warningsAreErrors || parsing.messages().AnyFatalError())) {
+    errs() << driver.prefix << "could not scan " << path << '\n';
+    parsing.messages().Emit(std::cerr, parsing.cooked());
+    exitStatus = EXIT_FAILURE;
+    return;
+  }
+  parsing.Parse(&std::cout);
+  parsing.messages().Emit(std::cerr, parsing.cooked());
+  if (!parsing.consumedWholeFile()) {
+    parsing.EmitMessage(std::cerr, parsing.finalRestingPlace(),
+                        "parser FAIL (final position)");
+    exitStatus = EXIT_FAILURE;
+    return;
+  }
+  if ((!parsing.messages().empty() &&
+       (driver.warningsAreErrors || parsing.messages().AnyFatalError())) ||
+      !parsing.parseTree().has_value()) {
+    errs() << driver.prefix << "could not parse " << path << '\n';
+    exitStatus = EXIT_FAILURE;
+    return;
+  }
+  auto &parseTree{*parsing.parseTree()};
+  Fortran::semantics::Semantics semantics{semanticsContext, parseTree,
+                                          parsing.cooked()};
+  semantics.Perform();
+  semantics.EmitMessages(std::cerr);
+  if (semantics.AnyFatalError()) {
+    errs() << driver.prefix << "semantic errors in " << path << '\n';
+    exitStatus = EXIT_FAILURE;
+    return;
+  }
+
+  // MLIR+FIR
+  fir::NameUniquer nameUniquer;
+  auto burnside = Br::BurnsideBridge::create(semanticsContext.defaultKinds(),
+                                             &parsing.cooked());
+  fir::KindMapping kindMap{&burnside.getMLIRContext()};
+  burnside.lower(parseTree, nameUniquer);
+  mlir::ModuleOp mlirModule = burnside.getModule();
+  mlir::PassManager pm = mlirModule.getContext();
+
+  pm.addPass(fir::createMemToRegPass());
+  pm.addPass(fir::createCSEPass());
+  pm.addPass(fir::createLowerToLoopPass());
+  pm.addPass(fir::createFIRToStdPass(kindMap));
+  pm.addPass(mlir::createLowerToCFGPass());
+
+  if (EmitLLVM) {
+    pm.addPass(fir::createFIRToLLVMPass(nameUniquer));
+    pm.addPass(fir::createLLVMDialectToLLVMPass(OutputFilename + ".ll"));
+  }
+
+  if (mlir::succeeded(pm.run(mlirModule))) {
+    std::error_code ec;
+    raw_fd_ostream out(OutputFilename, ec);
+    if (!ec)
+      mlirModule.print(out);
+  } else {
+    errs() << "oops, pass manager reported failure\n";
+  }
+}
+
+} // namespace
+
+int main(int argc, char **argv) {
+  InitLLVM y(argc, argv);
+  (void)y;
```

Comment:
Since we're using C++17 we can just mark this variable `[[maybe_unused]]` for clarity